### PR TITLE
Improve error messages when decoding

### PR DESCRIPTION
**Pull request by pnwpedro**
### Description

* Adds clearer error message when encountering an wire / codec mismatch while decoding
* Fixes missing support for `Stream` and `Bytes` tokens in `DynamicCodec`
* Enforce supported tokens in `ClassCodec.decode()`

Previous:
```
com.fauna.exception.ClientException: Unexpected token `START_OBJECT` decoding with `class com.fauna.codec.codecs.ClassCodec`
```

Now:
```
Unable to decode `Array` with `ClassCodec<Item>`. Supported types for codec are [Null, Object, Document, Ref].
```

### Motivation and context

BT-5095

A bug-basher encountered a situation where they modeled the response incorrectly. They said they were expecting a singular class when they were expecting an array. It proved very difficult to debug with the existing error message.

The reason we see START_OBJECT in the original error message rather than START_ARRAY was because we weren't enforcing the supported tokens specifically in the ClassCodec. This meant that we had START_ARRAY, then we advance the token stream and hit START_OBJECT, but the ClassCodec is expecting a FIELD_NAME token, at which point it was throwing because START_OBJECT != FIELD_NAME. This PR also fixes that issue.

### How was the change tested?

* Adds a complete test suite for this error across all codecs
* Adds a test to assert the DynamicCodec supports all modeled Fauna types

### Change types

* - [X] Bug fix (non-breaking change that fixes an issue)
* - [x] New feature (non-breaking change that adds functionality)
* - [ ] Breaking change (backwards-incompatible fix or feature)

### Checklist:

* - [X] My code follows the code style of this project.
* - [ ] My change requires a change to Fauna documentation.
* - [ ] My change requires a change to the README, and I have updated it accordingly.

----
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/fauna/codec/FaunaTokenType.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+    public FaunaTokenType getEndToken() throws IOException {
+        switch (this) {
+            case START_DOCUMENT: return END_DOCUMENT;
+            case START_OBJECT: return END_OBJECT;
+            case START_ARRAY: return END_ARRAY;
+            case START_PAGE: return END_PAGE;
+            case START_SET: return END_SET;
+            case START_REF: return END_REF;
+            default: throw new IOException("No end token for " + this.name());
```
Comment (erickpintor) [inline src/main/java/com/fauna/codec/FaunaTokenType.java:57]: Minor: this feels like an `IllegalStateException` to me. We're not doing any IO in here.

File: src/main/java/com/fauna/codec/FaunaTokenType.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        }
+    }
+
+    public FaunaType getFaunaType() {
+        switch (this) {
+            case START_OBJECT:
```
Comment (erickpintor) [inline src/main/java/com/fauna/codec/FaunaTokenType.java:63]: Q: wouldn't an `END_OBJECT` (and other variants) represent the same fauna type?

File: src/main/java/com/fauna/codec/FaunaType.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    Double,
+    String,
+    Date,
+    Time,
+    Boolean,
+    Object,
+    Ref,
+    Document,
+    Array,
+    Bytes, Null, Stream, Module, Set, Unknown
```
Comment (erickpintor) [inline src/main/java/com/fauna/codec/FaunaType.java:15]: Suggestion: `Unknown` is technically not a type. Removing it from here and making `getFaunaType()` return an `Optional` (or fail with an exception) would make this enum more precise.

File: src/main/java/com/fauna/codec/FaunaTokenType.java (hunk 1 above)
Comment (pnwpedro) [reply src/main/java/com/fauna/codec/FaunaTokenType.java:57]: Hmm, not sure how this ended up as IOException tbh. This was just moved, so I didn't touch the contents–will fix tho!

File: src/main/java/com/fauna/codec/FaunaTokenType.java (hunk 2 above)
Comment (pnwpedro) [reply src/main/java/com/fauna/codec/FaunaTokenType.java:63]: you know, it doesn't hurt, so i'll include them :shrug:

**Review by findgriffin — APPROVED** [review]

**Pull request merged**